File: server/game-engine/src/main/java/com/deepdame/engine/core/model/Piece.java
package com.deepdame.engine.core.model;

public record Piece(PieceType type, boolean isKing) {

    public Piece regular(PieceType type){
        return new Piece(type, false);
    }

    public Piece promote(){
        return new Piece(this.type, true);
    }
}

File: client/deepdame/lib/game-engine/model/piece.dart
import 'piece_type.dart';

class Piece {
  final PieceType type;
  final bool isKing;

  const Piece(this.type, this.isKing);

  // Static helper to create a regular piece
  static Piece regular(PieceType type) {
    return Piece(type, false);
  }

  // Returns a new Piece that is a King
  Piece promote() {
    return Piece(type, true);
  }
}